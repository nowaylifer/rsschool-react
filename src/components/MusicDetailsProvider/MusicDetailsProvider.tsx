import { useQueryParam, NumberParam } from 'use-query-params';
import {
  PropsWithChildren,
  createContext,
  useCallback,
  useEffect,
  useContext,
  useState,
} from 'react';
import musicApi from '../../services/musicApi';
import { getUpdatedQueryString } from '../../utils';
import { Album } from '../../types';

export interface MusicDetailsContext {
  getURLForAlbumDetails(albumId: number): string;
  albumDetails: Album | null;
}

const MusicDetailsContext = createContext<MusicDetailsContext | null>(null);

const MusicDetailsProvider = (props: PropsWithChildren) => {
  const [detailsParam] = useQueryParam('details', NumberParam);
  const [albumDetails, setAlbumDetails] = useState<Album | null>(null);

  useEffect(() => {
    if (!detailsParam) return;

    const fetchDetails = async () => {
      const result = await musicApi.fetchAlbumDetails(detailsParam);
      setAlbumDetails(result);
      console.log(result);
    };

    fetchDetails();
  }, [detailsParam]);

  const getURLForAlbumDetails = useCallback(
    (albumId: number) => getUpdatedQueryString({ details: albumId }),
    []
  );

  return (
    <MusicDetailsContext.Provider
      value={{ getURLForAlbumDetails, albumDetails } satisfies MusicDetailsContext}
      {...props}
    />
  );
};

export const useMusicDetails = () => {
  const contextValue = useContext(MusicDetailsContext);
  if (!contextValue) throw new Error('useMusicDetails must be used inside MusicDetailsContext');
  return contextValue;
};

export default MusicDetailsProvider;
