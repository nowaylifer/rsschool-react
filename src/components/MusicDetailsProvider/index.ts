export { default as MusicDetailsProvider, useMusicDetails } from './MusicDetailsProvider';
