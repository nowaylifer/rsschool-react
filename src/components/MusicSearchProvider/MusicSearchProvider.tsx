import {
  PropsWithChildren,
  createContext,
  useCallback,
  useEffect,
  useReducer,
  useContext,
} from 'react';
import musicApi, { SearchOptions } from '../../services/musicApi';
import { reducer, initialState } from './MusicSearchProvider.reducer';
import type { MusicSearchContext } from './MusicSearchProvider.types';
import { DEFAULT_PAGE, DEFAULT_PAGE_SIZE, PAGE_SIZES } from './MusicSearchProvider.config';
import { useQueryParams, NumberParam, StringParam, withDefault } from 'use-query-params';
import { getUpdatedQueryString } from '../../utils';

const MusicSearchContext = createContext<MusicSearchContext | null>(null);

const MusicSearchProvider = (props: PropsWithChildren) => {
  const [state, dispatch] = useReducer(reducer, initialState);
  const [queryParams, setQueryParams] = useQueryParams({
    q: withDefault(StringParam, null),
    page: withDefault(NumberParam, DEFAULT_PAGE),
    pageSize: withDefault(NumberParam, DEFAULT_PAGE_SIZE),
  });

  useEffect(() => {
    if (queryParams.q === null) return;

    const searchMusic = async () => {
      dispatch({ type: 'START_SEARCH' });

      const { pageSize, page } = queryParams;

      const options: SearchOptions = {
        limit: pageSize,
        index: (page - 1) * pageSize,
      };

      try {
        const result = await musicApi.search(queryParams.q!, options);
        dispatch({ type: 'SEARCH_RESOLVED', payload: result });
      } catch (error) {
        dispatch({
          type: 'REJECTED',
          payload:
            error instanceof Error ? error : new Error(`Unknown error: ${error?.toString()}`),
        });
      }
    };

    searchMusic();
  }, [queryParams.page, queryParams.pageSize, queryParams.q]);

  const submitSearch = useCallback((query: string) => {
    setQueryParams({
      q: query,
      page: DEFAULT_PAGE,
    });
  }, []);

  const changePageSize = useCallback((size: number) => {
    setQueryParams({ page: DEFAULT_PAGE, pageSize: size });
  }, []);

  const getURLForPage = useCallback((page: number) => {
    const param = { page: page === DEFAULT_PAGE ? undefined : page };
    return getUpdatedQueryString(param);
  }, []);

  return (
    <MusicSearchContext.Provider
      value={{
        submitSearch,
        getURLForPage,
        changePageSize,
        queryParams,
        albums: state.albums,
        error: state.error,
        loading: state.status === 'pending',
        totalItems: state.totalItems,
        pageSizes: PAGE_SIZES,
      }}
      {...props}
    />
  );
};

export const useMusicSearch = () => {
  const contextValue = useContext(MusicSearchContext);
  if (!contextValue) throw new Error('useMusicSearch must be used inside MusicSearchContext');
  return contextValue;
};

export default MusicSearchProvider;
