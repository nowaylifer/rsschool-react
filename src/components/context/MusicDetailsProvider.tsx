import { useQueryParam, NumberParam } from 'use-query-params';
import {
  PropsWithChildren,
  createContext,
  useCallback,
  useEffect,
  useContext,
  useState,
} from 'react';
import musicApi from '../../services/musicApi';
import { getUpdatedQueryString } from '../../utils';
import { Album } from '../../types';

type DetailsStatus = 'idle' | 'loading' | 'resolved' | 'rejected';

export interface MusicDetailsContext {
  getURLForAlbumDetails(albumId: number): string;
  unsetDetails(): void;
  albumDetails: Album | null;
  status: DetailsStatus;
}

const MusicDetailsContext = createContext<MusicDetailsContext | null>(null);

const MusicDetailsProvider = (props: PropsWithChildren) => {
  const [detailsParam, setDetailsParam] = useQueryParam('details', NumberParam);
  const [albumDetails, setAlbumDetails] = useState<Album | null>(null);
  const [status, setStatus] = useState<DetailsStatus>('idle');

  useEffect(() => {
    if (!detailsParam) return;

    const fetchDetails = async () => {
      setStatus('loading');
      const result = await musicApi.fetchAlbumDetails(detailsParam);
      setStatus('resolved');
      setAlbumDetails(result);
      console.log(result);
    };

    fetchDetails();
  }, [detailsParam]);

  const getURLForAlbumDetails = useCallback(
    (albumId: number) => getUpdatedQueryString({ details: albumId }),
    []
  );

  const unsetDetails = useCallback(() => {
    setStatus('idle');
    setAlbumDetails(null);
    setDetailsParam(undefined);
  }, []);

  return (
    <MusicDetailsContext.Provider
      value={{ getURLForAlbumDetails, albumDetails, unsetDetails, status }}
      {...props}
    />
  );
};

export const useMusicDetails = () => {
  const contextValue = useContext(MusicDetailsContext);
  if (!contextValue) throw new Error('useMusicDetails must be used inside MusicDetailsContext');
  return contextValue;
};

export default MusicDetailsProvider;
