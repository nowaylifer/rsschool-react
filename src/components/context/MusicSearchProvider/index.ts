export { default, useMusicSearch } from './MusicSearchProvider';
