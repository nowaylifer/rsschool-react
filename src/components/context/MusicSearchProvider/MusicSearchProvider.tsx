import {
  PropsWithChildren,
  createContext,
  useCallback,
  useEffect,
  useReducer,
  useContext,
} from 'react';
import musicApi, { SearchOptions } from '../../../services/musicApi';
import { reducer, initialState } from './MusicSearchProvider.reducer';
import type { MusicSearchContext } from './MusicSearchProvider.types';
import { DEFAULT_PAGE, DEFAULT_PAGE_SIZE, PAGE_SIZES } from './MusicSearchProvider.config';
import { useQueryParams, NumberParam, StringParam, withDefault } from 'use-query-params';

const MusicSearchContext = createContext<MusicSearchContext | null>(null);

const MusicSearchProvider = (props: PropsWithChildren) => {
  const [state, dispatch] = useReducer(reducer, initialState);
  const [queryParams, setQueryParams] = useQueryParams({
    q: withDefault(StringParam, ''),
    page: withDefault(NumberParam, DEFAULT_PAGE),
    pageSize: withDefault(NumberParam, DEFAULT_PAGE_SIZE),
  });

  useEffect(() => {
    const searchMusic = async () => {
      dispatch({ type: 'START_SEARCH' });

      const { pageSize, page, q } = queryParams;

      const options: SearchOptions<'album'> = {
        type: 'album',
        limit: pageSize,
        index: (page - 1) * pageSize,
      };

      try {
        const query = q.trim();
        let result;

        if (query) {
          result = await musicApi.search(query, options);
        } else {
          result = await musicApi.fetchEditorialReleases(options);
        }

        dispatch({ type: 'SEARCH_RESOLVED', payload: result });
      } catch (error) {
        dispatch({
          type: 'REJECTED',
          payload:
            error instanceof Error ? error : new Error(`Unknown error: ${error?.toString()}`),
        });
      }
    };

    searchMusic();
  }, [queryParams.page, queryParams.pageSize, queryParams.q]);

  const submitSearch = useCallback((query: string) => {
    setQueryParams({
      q: query,
      page: DEFAULT_PAGE,
    });
  }, []);

  const changePageSize = useCallback((size: number) => {
    setQueryParams({ page: DEFAULT_PAGE, pageSize: size });
  }, []);

  const changePage = useCallback((page: number) => {
    setQueryParams({ page });
  }, []);

  return (
    <MusicSearchContext.Provider
      value={{
        submitSearch,
        changePage,
        changePageSize,
        queryParams,
        albums: state.albums,
        error: state.error,
        loading: state.status === 'pending',
        totalItems: state.totalItems,
        pageSizes: PAGE_SIZES,
      }}
      {...props}
    />
  );
};

export const useMusicSearch = () => {
  const contextValue = useContext(MusicSearchContext);
  if (!contextValue) throw new Error('useMusicSearch must be used inside MusicSearchContext');
  return contextValue;
};

export default MusicSearchProvider;
