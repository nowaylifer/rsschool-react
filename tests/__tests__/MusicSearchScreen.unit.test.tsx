import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MemoryRouter } from 'react-router-dom';
import { mockMusicDetailsContextValue } from '../__mocks__/MusicDetailsContext';
import {
  mockMusicSearchContextValueError,
  mockMusicSearchContextValueSuccess,
} from '../__mocks__/MusicSearchContext';
import MusicSearchScreen from '../../src/screens/MusicSearchScreen';
import { MusicDetailsContext } from '../../src/context/MusicDetailsProvider';
import { MusicSearchContext, MusicSearchContextType } from '../../src/context/MusicSearchProvider';
// import { TEST_API_URL } from '../test-utils';

jest.mock('../../src/hooks/useMediaQuery.ts', () => ({
  __esModule: true,
  default: jest.fn(() => false),
}));

jest.mock('react', () => ({
  __esModule: true,
  ...jest.requireActual('react'),
  useId: jest.fn(() => 'id123'),
}));

const customRender = (musicSearchContextValue: MusicSearchContextType) => {
  render(
    <MemoryRouter>
      <MusicSearchContext.Provider value={musicSearchContextValue}>
        <MusicDetailsContext.Provider value={mockMusicDetailsContextValue}>
          <MusicSearchScreen />
        </MusicDetailsContext.Provider>
      </MusicSearchContext.Provider>
    </MemoryRouter>
  );
};

describe('MusicSearchScreen', () => {
  describe('when data is successfully loaded', () => {
    beforeEach(() => customRender(mockMusicSearchContextValueSuccess));

    it('renders cards with specified data', () => {
      expect(screen.getAllByTestId('card')).toHaveLength(
        mockMusicSearchContextValueSuccess.albums.length
      );

      mockMusicSearchContextValueSuccess.albums.forEach((item) => {
        expect(screen.getByText(item.title)).toBeInTheDocument();
      });
    });

    it('renders search form that could be submited', async () => {
      const searchInput = screen.getByRole<HTMLInputElement>('textbox', { name: /search music/i });
      expect(searchInput).toHaveValue(mockMusicSearchContextValueSuccess.queryParams.q);

      searchInput.value = '';
      const newQuery = 'nirvana';

      await userEvent.type(searchInput, newQuery);
      await userEvent.click(screen.getByRole('button', { name: /search/i }));

      expect(mockMusicSearchContextValueSuccess.submitSearch).toHaveBeenCalledTimes(1);
      expect(mockMusicSearchContextValueSuccess.submitSearch).toHaveBeenCalledWith(newQuery);
    });

    it('matches snapshot', () => {
      expect(document.body).toMatchSnapshot();
    });
  });

  describe('when there is an error loading data', () => {
    beforeEach(() => customRender(mockMusicSearchContextValueError));

    it('renders error message', () => {
      expect(screen.getByText(mockMusicSearchContextValueError.error!.message)).toBeInTheDocument();
    });

    it('matches snapshot', () => {
      expect(document.body).toMatchSnapshot();
    });
  });
});
