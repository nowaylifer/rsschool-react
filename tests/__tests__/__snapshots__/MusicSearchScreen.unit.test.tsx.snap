// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MusicSearchScreen when data is successfully loaded matches snapshot 1`] = `
<body>
  <div>
    <main>
      <form
        class="mb-10"
        role="search"
      >
        <label
          class="sr-only"
          for="id123"
        >
          Search music
        </label>
        <input
          class="w-full appearance-none rounded-md border px-3 py-2 font-medium leading-tight text-gray-700 shadow focus:outline-none focus:ring mb-3"
          id="id123"
          name="search"
          placeholder="Search"
          value="eminem"
        />
        <button
          class="rounded bg-blue-500 px-4 py-2 font-semibold text-white hover:bg-blue-600 focus:outline-none focus:ring"
        >
          Search
        </button>
      </form>
      <div
        class="flex flex-col items-center justify-center gap-x-2 gap-y-4 md:flex-row mb-10"
      >
        <nav
          class=""
          data-testid="pagination"
        >
          <ul
            class="flex justify-center"
          >
            <button
              aria-label="Previous page"
              class="border-blue-gray-100 mx-1 flex h-9 w-9 items-center justify-center rounded-full border p-0 text-sm font-semibold transition duration-150 ease-in-out hover:bg-gray-100 focus:delay-75 pointer-events-none bg-gray-50 text-gray-400 focus:outline-none focus:ring-0"
            >
              <div />
            </button>
            <button
              aria-label="Page 1"
              class="border-blue-gray-100 mx-1 flex h-9 w-9 items-center justify-center rounded-full border p-0 text-sm font-semibold transition duration-150 ease-in-out focus:outline-none focus:ring focus:delay-75 bg-pink-500 text-white shadow-md hover:bg-pink-500"
            >
              1
            </button>
            <button
              aria-label="Page 2"
              class="border-blue-gray-100 text-blue-gray-500 mx-1 flex h-9 w-9 items-center justify-center rounded-full border bg-white p-0 text-sm font-semibold transition duration-150 ease-in-out hover:bg-gray-100 focus:outline-none focus:ring focus:delay-75"
            >
              2
            </button>
            <button
              aria-label="Page 3"
              class="border-blue-gray-100 text-blue-gray-500 mx-1 flex h-9 w-9 items-center justify-center rounded-full border bg-white p-0 text-sm font-semibold transition duration-150 ease-in-out hover:bg-gray-100 focus:outline-none focus:ring focus:delay-75"
            >
              3
            </button>
            <button
              aria-label="Page 4"
              class="border-blue-gray-100 text-blue-gray-500 mx-1 flex h-9 w-9 items-center justify-center rounded-full border bg-white p-0 text-sm font-semibold transition duration-150 ease-in-out hover:bg-gray-100 focus:outline-none focus:ring focus:delay-75"
            >
              4
            </button>
            <button
              aria-label="Page 5"
              class="border-blue-gray-100 text-blue-gray-500 mx-1 flex h-9 w-9 items-center justify-center rounded-full border bg-white p-0 text-sm font-semibold transition duration-150 ease-in-out hover:bg-gray-100 focus:outline-none focus:ring focus:delay-75"
            >
              5
            </button>
            <button
              aria-label="Page 6"
              class="border-blue-gray-100 text-blue-gray-500 mx-1 flex h-9 w-9 items-center justify-center rounded-full border bg-white p-0 text-sm font-semibold transition duration-150 ease-in-out hover:bg-gray-100 focus:outline-none focus:ring focus:delay-75"
            >
              6
            </button>
            <button
              aria-label="Page 7"
              class="border-blue-gray-100 text-blue-gray-500 mx-1 flex h-9 w-9 items-center justify-center rounded-full border bg-white p-0 text-sm font-semibold transition duration-150 ease-in-out hover:bg-gray-100 focus:outline-none focus:ring focus:delay-75"
            >
              7
            </button>
            <button
              class="border-blue-gray-100 text-blue-gray-500 mx-1 flex h-9 w-9 items-center justify-center rounded-full border bg-white p-0 text-sm font-semibold transition duration-150 ease-in-out hover:bg-gray-100 focus:outline-none focus:ring focus:delay-75"
            >
              ...
            </button>
            <button
              aria-label="Page 28"
              class="border-blue-gray-100 text-blue-gray-500 mx-1 flex h-9 w-9 items-center justify-center rounded-full border bg-white p-0 text-sm font-semibold transition duration-150 ease-in-out hover:bg-gray-100 focus:outline-none focus:ring focus:delay-75"
            >
              28
            </button>
            <button
              aria-label="Next page"
              class="border-blue-gray-100 text-blue-gray-500 mx-1 flex h-9 w-9 items-center justify-center rounded-full border bg-white p-0 text-sm font-semibold transition duration-150 ease-in-out hover:bg-gray-100 focus:outline-none focus:ring focus:delay-75"
            >
              <div
                class="rotate-180"
              />
            </button>
          </ul>
        </nav>
        <select
          aria-label="Select page size"
          class="border-blue-gray-100 cursor-pointer rounded-full border bg-white px-2 py-2 text-sm font-semibold transition focus:outline-none focus:ring [&:not(:focus)]:hover:bg-gray-100"
        >
          <option
            value="10"
          >
            10 / page
          </option>
          <option
            value="20"
          >
            20 / page
          </option>
          <option
            value="30"
          >
            30 / page
          </option>
        </select>
      </div>
      <div
        class="grid grid-cols-[repeat(auto-fit,minmax(16rem,1fr))] justify-items-center gap-6"
      >
        <div
          class="flex max-w-[22rem]"
        >
          <a
            data-testid="card"
            href="/"
          >
            <div
              class="relative flex flex-col items-center gap-8 rounded-xl bg-white bg-clip-border px-6 py-6 text-gray-700 shadow transition-shadow hover:shadow-lg"
            >
              <div
                class="relative overflow-hidden rounded-xl bg-white bg-clip-border shadow-lg"
              >
                <img
                  class="transition-[filter]"
                  height="500"
                  src="https://e-cdns-images.dzcdn.net/images/cover/ec3c8ed67427064c70f67e5815b74cef/500x500-000000-80-0-0.jpg"
                  width="500"
                />
              </div>
              <div
                class="text-center"
              >
                <h4
                  class="mb-2 text-lg font-semibold leading-snug tracking-normal text-gray-900 antialiased"
                >
                  The Eminem Show
                </h4>
                <p
                  class="leading-relax font-medium antialiased text-gray-500"
                >
                  Eminem
                </p>
              </div>
            </div>
          </a>
        </div>
        <div
          class="flex max-w-[22rem]"
        >
          <a
            data-testid="card"
            href="/"
          >
            <div
              class="relative flex flex-col items-center gap-8 rounded-xl bg-white bg-clip-border px-6 py-6 text-gray-700 shadow transition-shadow hover:shadow-lg"
            >
              <div
                class="relative overflow-hidden rounded-xl bg-white bg-clip-border shadow-lg"
              >
                <img
                  class="transition-[filter]"
                  height="500"
                  src="https://e-cdns-images.dzcdn.net/images/cover/941c2d3c366affdc662956559e078a4e/500x500-000000-80-0-0.jpg"
                  width="500"
                />
              </div>
              <div
                class="text-center"
              >
                <h4
                  class="mb-2 text-lg font-semibold leading-snug tracking-normal text-gray-900 antialiased"
                >
                  The Marshall Mathers LP
                </h4>
                <p
                  class="leading-relax font-medium antialiased text-gray-500"
                >
                  Eminem
                </p>
              </div>
            </div>
          </a>
        </div>
        <div
          class="flex max-w-[22rem]"
        >
          <a
            data-testid="card"
            href="/"
          >
            <div
              class="relative flex flex-col items-center gap-8 rounded-xl bg-white bg-clip-border px-6 py-6 text-gray-700 shadow transition-shadow hover:shadow-lg"
            >
              <div
                class="relative overflow-hidden rounded-xl bg-white bg-clip-border shadow-lg"
              >
                <img
                  class="transition-[filter]"
                  height="500"
                  src="https://e-cdns-images.dzcdn.net/images/cover/be682506145061814eddee648edb7c59/500x500-000000-80-0-0.jpg"
                  width="500"
                />
              </div>
              <div
                class="text-center"
              >
                <h4
                  class="mb-2 text-lg font-semibold leading-snug tracking-normal text-gray-900 antialiased"
                >
                  Recovery
                </h4>
                <p
                  class="leading-relax font-medium antialiased text-gray-500"
                >
                  Eminem
                </p>
              </div>
            </div>
          </a>
        </div>
        <div
          class="flex max-w-[22rem]"
        >
          <a
            data-testid="card"
            href="/"
          >
            <div
              class="relative flex flex-col items-center gap-8 rounded-xl bg-white bg-clip-border px-6 py-6 text-gray-700 shadow transition-shadow hover:shadow-lg"
            >
              <div
                class="relative overflow-hidden rounded-xl bg-white bg-clip-border shadow-lg"
              >
                <img
                  class="transition-[filter]"
                  height="500"
                  src="https://e-cdns-images.dzcdn.net/images/cover/5026c46b38ed035092df6cfc5a4a288f/500x500-000000-80-0-0.jpg"
                  width="500"
                />
              </div>
              <div
                class="text-center"
              >
                <h4
                  class="mb-2 text-lg font-semibold leading-snug tracking-normal text-gray-900 antialiased"
                >
                  Encore
                </h4>
                <p
                  class="leading-relax font-medium antialiased text-gray-500"
                >
                  Eminem
                </p>
              </div>
            </div>
          </a>
        </div>
        <div
          class="flex max-w-[22rem]"
        >
          <a
            data-testid="card"
            href="/"
          >
            <div
              class="relative flex flex-col items-center gap-8 rounded-xl bg-white bg-clip-border px-6 py-6 text-gray-700 shadow transition-shadow hover:shadow-lg"
            >
              <div
                class="relative overflow-hidden rounded-xl bg-white bg-clip-border shadow-lg"
              >
                <img
                  class="transition-[filter]"
                  height="500"
                  src="https://e-cdns-images.dzcdn.net/images/cover/bf74fc764097630ba58782ae79cfbee6/500x500-000000-80-0-0.jpg"
                  width="500"
                />
              </div>
              <div
                class="text-center"
              >
                <h4
                  class="mb-2 text-lg font-semibold leading-snug tracking-normal text-gray-900 antialiased"
                >
                  Kamikaze
                </h4>
                <p
                  class="leading-relax font-medium antialiased text-gray-500"
                >
                  Eminem
                </p>
              </div>
            </div>
          </a>
        </div>
        <div
          class="flex max-w-[22rem]"
        >
          <a
            data-testid="card"
            href="/"
          >
            <div
              class="relative flex flex-col items-center gap-8 rounded-xl bg-white bg-clip-border px-6 py-6 text-gray-700 shadow transition-shadow hover:shadow-lg"
            >
              <div
                class="relative overflow-hidden rounded-xl bg-white bg-clip-border shadow-lg"
              >
                <img
                  class="transition-[filter]"
                  height="500"
                  src="https://e-cdns-images.dzcdn.net/images/cover/7c71224818ac99b16570357b2f98e041/500x500-000000-80-0-0.jpg"
                  width="500"
                />
              </div>
              <div
                class="text-center"
              >
                <h4
                  class="mb-2 text-lg font-semibold leading-snug tracking-normal text-gray-900 antialiased"
                >
                  The Marshall Mathers LP2 (Deluxe)
                </h4>
                <p
                  class="leading-relax font-medium antialiased text-gray-500"
                >
                  Eminem
                </p>
              </div>
            </div>
          </a>
        </div>
        <div
          class="flex max-w-[22rem]"
        >
          <a
            data-testid="card"
            href="/"
          >
            <div
              class="relative flex flex-col items-center gap-8 rounded-xl bg-white bg-clip-border px-6 py-6 text-gray-700 shadow transition-shadow hover:shadow-lg"
            >
              <div
                class="relative overflow-hidden rounded-xl bg-white bg-clip-border shadow-lg"
              >
                <img
                  class="transition-[filter]"
                  height="500"
                  src="https://e-cdns-images.dzcdn.net/images/cover/cac1e3634d43bdbf0ab03b6f36ecaf7d/500x500-000000-80-0-0.jpg"
                  width="500"
                />
              </div>
              <div
                class="text-center"
              >
                <h4
                  class="mb-2 text-lg font-semibold leading-snug tracking-normal text-gray-900 antialiased"
                >
                  FEMINEM
                </h4>
                <p
                  class="leading-relax font-medium antialiased text-gray-500"
                >
                  Antifuchs
                </p>
              </div>
            </div>
          </a>
        </div>
        <div
          class="flex max-w-[22rem]"
        >
          <a
            data-testid="card"
            href="/"
          >
            <div
              class="relative flex flex-col items-center gap-8 rounded-xl bg-white bg-clip-border px-6 py-6 text-gray-700 shadow transition-shadow hover:shadow-lg"
            >
              <div
                class="relative overflow-hidden rounded-xl bg-white bg-clip-border shadow-lg"
              >
                <img
                  class="transition-[filter]"
                  height="500"
                  src="https://e-cdns-images.dzcdn.net/images/cover/f0b8f9a786d96d8f9fc2a765462c17a6/500x500-000000-80-0-0.jpg"
                  width="500"
                />
              </div>
              <div
                class="text-center"
              >
                <h4
                  class="mb-2 text-lg font-semibold leading-snug tracking-normal text-gray-900 antialiased"
                >
                  Encore (Deluxe Version)
                </h4>
                <p
                  class="leading-relax font-medium antialiased text-gray-500"
                >
                  Eminem
                </p>
              </div>
            </div>
          </a>
        </div>
        <div
          class="flex max-w-[22rem]"
        >
          <a
            data-testid="card"
            href="/"
          >
            <div
              class="relative flex flex-col items-center gap-8 rounded-xl bg-white bg-clip-border px-6 py-6 text-gray-700 shadow transition-shadow hover:shadow-lg"
            >
              <div
                class="relative overflow-hidden rounded-xl bg-white bg-clip-border shadow-lg"
              >
                <img
                  class="transition-[filter]"
                  height="500"
                  src="https://e-cdns-images.dzcdn.net/images/cover/1dbb5a67a722a9defa1d9d23b474cee7/500x500-000000-80-0-0.jpg"
                  width="500"
                />
              </div>
              <div
                class="text-center"
              >
                <h4
                  class="mb-2 text-lg font-semibold leading-snug tracking-normal text-gray-900 antialiased"
                >
                  Feminem
                </h4>
                <p
                  class="leading-relax font-medium antialiased text-gray-500"
                >
                  Justina Valentine
                </p>
              </div>
            </div>
          </a>
        </div>
        <div
          class="flex max-w-[22rem]"
        >
          <a
            data-testid="card"
            href="/"
          >
            <div
              class="relative flex flex-col items-center gap-8 rounded-xl bg-white bg-clip-border px-6 py-6 text-gray-700 shadow transition-shadow hover:shadow-lg"
            >
              <div
                class="relative overflow-hidden rounded-xl bg-white bg-clip-border shadow-lg"
              >
                <img
                  class="transition-[filter]"
                  height="500"
                  src="https://e-cdns-images.dzcdn.net/images/cover/55d424a53904d145b17869ba133843d7/500x500-000000-80-0-0.jpg"
                  width="500"
                />
              </div>
              <div
                class="text-center"
              >
                <h4
                  class="mb-2 text-lg font-semibold leading-snug tracking-normal text-gray-900 antialiased"
                >
                  Music To Be Murdered By - Side B (Deluxe Edition)
                </h4>
                <p
                  class="leading-relax font-medium antialiased text-gray-500"
                >
                  Eminem
                </p>
              </div>
            </div>
          </a>
        </div>
      </div>
      <div
        class="flex flex-col items-center justify-center gap-x-2 gap-y-4 md:flex-row mt-10"
      >
        <nav
          class=""
          data-testid="pagination"
        >
          <ul
            class="flex justify-center"
          >
            <button
              aria-label="Previous page"
              class="border-blue-gray-100 mx-1 flex h-9 w-9 items-center justify-center rounded-full border p-0 text-sm font-semibold transition duration-150 ease-in-out hover:bg-gray-100 focus:delay-75 pointer-events-none bg-gray-50 text-gray-400 focus:outline-none focus:ring-0"
            >
              <div />
            </button>
            <button
              aria-label="Page 1"
              class="border-blue-gray-100 mx-1 flex h-9 w-9 items-center justify-center rounded-full border p-0 text-sm font-semibold transition duration-150 ease-in-out focus:outline-none focus:ring focus:delay-75 bg-pink-500 text-white shadow-md hover:bg-pink-500"
            >
              1
            </button>
            <button
              aria-label="Page 2"
              class="border-blue-gray-100 text-blue-gray-500 mx-1 flex h-9 w-9 items-center justify-center rounded-full border bg-white p-0 text-sm font-semibold transition duration-150 ease-in-out hover:bg-gray-100 focus:outline-none focus:ring focus:delay-75"
            >
              2
            </button>
            <button
              aria-label="Page 3"
              class="border-blue-gray-100 text-blue-gray-500 mx-1 flex h-9 w-9 items-center justify-center rounded-full border bg-white p-0 text-sm font-semibold transition duration-150 ease-in-out hover:bg-gray-100 focus:outline-none focus:ring focus:delay-75"
            >
              3
            </button>
            <button
              aria-label="Page 4"
              class="border-blue-gray-100 text-blue-gray-500 mx-1 flex h-9 w-9 items-center justify-center rounded-full border bg-white p-0 text-sm font-semibold transition duration-150 ease-in-out hover:bg-gray-100 focus:outline-none focus:ring focus:delay-75"
            >
              4
            </button>
            <button
              aria-label="Page 5"
              class="border-blue-gray-100 text-blue-gray-500 mx-1 flex h-9 w-9 items-center justify-center rounded-full border bg-white p-0 text-sm font-semibold transition duration-150 ease-in-out hover:bg-gray-100 focus:outline-none focus:ring focus:delay-75"
            >
              5
            </button>
            <button
              aria-label="Page 6"
              class="border-blue-gray-100 text-blue-gray-500 mx-1 flex h-9 w-9 items-center justify-center rounded-full border bg-white p-0 text-sm font-semibold transition duration-150 ease-in-out hover:bg-gray-100 focus:outline-none focus:ring focus:delay-75"
            >
              6
            </button>
            <button
              aria-label="Page 7"
              class="border-blue-gray-100 text-blue-gray-500 mx-1 flex h-9 w-9 items-center justify-center rounded-full border bg-white p-0 text-sm font-semibold transition duration-150 ease-in-out hover:bg-gray-100 focus:outline-none focus:ring focus:delay-75"
            >
              7
            </button>
            <button
              class="border-blue-gray-100 text-blue-gray-500 mx-1 flex h-9 w-9 items-center justify-center rounded-full border bg-white p-0 text-sm font-semibold transition duration-150 ease-in-out hover:bg-gray-100 focus:outline-none focus:ring focus:delay-75"
            >
              ...
            </button>
            <button
              aria-label="Page 28"
              class="border-blue-gray-100 text-blue-gray-500 mx-1 flex h-9 w-9 items-center justify-center rounded-full border bg-white p-0 text-sm font-semibold transition duration-150 ease-in-out hover:bg-gray-100 focus:outline-none focus:ring focus:delay-75"
            >
              28
            </button>
            <button
              aria-label="Next page"
              class="border-blue-gray-100 text-blue-gray-500 mx-1 flex h-9 w-9 items-center justify-center rounded-full border bg-white p-0 text-sm font-semibold transition duration-150 ease-in-out hover:bg-gray-100 focus:outline-none focus:ring focus:delay-75"
            >
              <div
                class="rotate-180"
              />
            </button>
          </ul>
        </nav>
        <select
          aria-label="Select page size"
          class="border-blue-gray-100 cursor-pointer rounded-full border bg-white px-2 py-2 text-sm font-semibold transition focus:outline-none focus:ring [&:not(:focus)]:hover:bg-gray-100"
        >
          <option
            value="10"
          >
            10 / page
          </option>
          <option
            value="20"
          >
            20 / page
          </option>
          <option
            value="30"
          >
            30 / page
          </option>
        </select>
      </div>
    </main>
  </div>
</body>
`;

exports[`MusicSearchScreen when there is an error loading data matches snapshot 1`] = `
<body>
  <div>
    <main>
      <form
        class="mb-10"
        role="search"
      >
        <label
          class="sr-only"
          for="id123"
        >
          Search music
        </label>
        <input
          class="w-full appearance-none rounded-md border px-3 py-2 font-medium leading-tight text-gray-700 shadow focus:outline-none focus:ring mb-3"
          id="id123"
          name="search"
          placeholder="Search"
          value="eminem"
        />
        <button
          class="rounded bg-blue-500 px-4 py-2 font-semibold text-white hover:bg-blue-600 focus:outline-none focus:ring"
        >
          Search
        </button>
      </form>
      <div
        class="flex flex-col items-center justify-center gap-4"
      >
        <div
          class="h-14 fill-slate-500"
        />
        <p
          class="text-xl font-semibold text-slate-600"
        >
          No result found
        </p>
      </div>
    </main>
  </div>
</body>
`;
